hydromt/data_source/data_source.py

```diff
@@ -0,0 +1,109 @@
+"""Abstract DataSource class."""
+from os.path import abspath, join
+from pathlib import Path
+from typing import Any, Callable, ClassVar, Dict, Optional, Union
+
+from pydantic import (
+    BaseModel,
+    Field,
+    model_validator,
+)
+
+from hydromt._typing import DataType
+from hydromt.data_adapter.caching import _uri_validator
+from hydromt.data_adapter.data_adapter_base import DataAdapterBase
+from hydromt.data_adapter.harmonization_settings import HarmonizationSettings
+from hydromt.driver import BaseDriver
+
+
+class DataSource(BaseModel):
+    """
+    A DataSource is a parsed section of a DataCatalog.
+
+    The DataSource, specific for a data type within HydroMT, is responsible for
+    validating the input from the DataCatalog, to
+    ensure the workflow fails as early as possible. A DataSource has information on
+    the driver that the data should be read with, and is responsible for initializing
+    this driver.
+    """
+
+    name: str
+    uri: str
+    data_adapter: DataAdapterBase
+    driver: BaseDriver
+    data_type: ClassVar[DataType]
+    root: Optional[str] = Field(default=None)
+    harmonization: HarmonizationSettings = Field(default_factory=HarmonizationSettings)
+    version: Optional[str] = Field(default=None)
+    provider: Optional[str] = Field(default=None)
+    driver_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    resolver_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    crs: Optional[int] = Field(default=None)
+
+    def summary(self) -> Dict[str, Any]:
+        """Return a summary of the DataSource."""
+        summ: Dict[str, Any] = self.model_dump(include={"uri"})
+        summ.update(
+            {
+                "data_type": self.__class__.data_type,
+                "driver": self.driver.__repr_name__(),
+                **self.data_adapter.harmonization_settings.meta,
+            }
+        )
+        return summ
+
+    # TODO: def to_file(self, **query_params) https://github.com/Deltares/hydromt/issues/840
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def _init_subclass_data_type(cls, data: Any, handler: Callable):
```

Comment:
it is not clear to me what this handler is supposed to do, can we elaborate on that?